File: 백준/Gold/2448. 별 찍기 － 11/별 찍기 － 11.java
import java.util.*;

public class Main {
    static int logN;
    static ArrayList<String> base;

    static void tri(int cnt, int n) {
        if (n > logN) {
            return;
        }

        if (n == 0) {
            for (String i : base) {
                String blank = " ".repeat(cnt);
                System.out.println(blank + i + blank);
                cnt -= 1;
            }
        } else {
            int length = base.size();
            int start = base.get(length-1).length();

            for (int i = 0; i < length; i++) {
                base.add(base.get(i) + " ".repeat(start) + base.get(i));
                start -= 2;
            }
            for (int i = length; i < base.size(); i++) {
                String blank = " ".repeat(cnt);
                System.out.println(blank + base.get(i) + blank);
                cnt -= 1;
            }
        }

        tri(cnt, n+1);

    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        logN = (int) (Math.log(n/3) / Math.log(2));
        base = new ArrayList<>(Arrays.asList("*", "* *", "*****"));

        tri(n-1, 0);
    }
}

File: 백준/Gold/2448. 별 찍기 － 11/별 찍기 － 11.py
# 별 찍기 - 11
import sys
input = sys.stdin.readline
import math

def tri(cnt, n):
    if n > logN:
        return

    if n == 0:
        for i in base:
            blank = cnt * " "
            print(blank + i + blank, sep="")
            cnt -= 1
    else:
        length = len(base)
        start = len(base[-1])
        for i in range(length):
            base.append(base[i] + " " * start + base[i])
            start -= 2

        for i in range(length, len(base)):
            blank = cnt * " "
            print(blank + base[i] + blank, sep="")
            cnt -= 1
            
    tri(cnt, n+1)


n = int(input())
logN = int(math.log(n//3, 2))
base = ["*", "* *", "*****"]

tri(n-1, 0)
